src/sys/fuchsia/selector.rs

```diff
@@ -0,0 +1,355 @@
+            .or_else(|e| if e.kind() == io::ErrorKind::NotFound {
+                Ok(())
+            } else {
+                Err(e)
+            })
+    }
+
+    pub ( in sys::fuchsia) fn register_handle<H>(&self,
+                                                 handle: &H,
+                                                 token: Token,
+                                                 interests: Ready,
+                                                 poll_opts: PollOpt) -> io::Result<()>
+        where H: magenta::HandleBase
+    {
+        if poll_opts.is_level() && !poll_opts.is_oneshot() {
+            panic!("Repeated level-triggered events are not supported on Fuchsia handles.");
+        }
+
+        handle.wait_async(&self.port,
+                          key_from_token_and_type(token, RegType::Handle),
+                          ready_to_signals(interests),
+                          poll_opts_to_wait_async(poll_opts))
+              .map_err(status_to_io_err)
+    }
+
+
+    pub ( in sys::fuchsia) fn deregister_handle<H>(&self,
+                                                   handle: &H,
+                                                   token: Token) -> io::Result<()>
+        where H: magenta::HandleBase
+    {
+        self.port.cancel(handle,
+                         key_from_token_and_type(token, RegType::Handle))
+                 .map_err(status_to_io_err)
+    }
+}
+
+fn ready_to_signals(ready: Ready) -> magenta::Signals {
+    // Get empty notifications for peer closing or other miscellaneous signals:
+    magenta_sys::MX_OBJECT_PEER_CLOSED |
+    magenta::MX_EVENT_SIGNALED |
+
+    if ready.is_writable() {
+        magenta_sys::MX_OBJECT_WRITABLE
+    } else {
+        magenta::MX_SIGNAL_NONE
+    } |
+
+    if ready.is_readable() {
+        magenta_sys::MX_OBJECT_READABLE
+    } else {
+        magenta::MX_SIGNAL_NONE
+    }
+}
+
+fn signals_to_ready(signals: magenta::Signals) -> Ready {
+    (if signals.contains(magenta_sys::MX_OBJECT_WRITABLE) {
+        Ready::writable()
+    } else {
+        Ready::empty()
+    }) |
+
+    (if signals.contains(magenta_sys::MX_OBJECT_READABLE) {
+        Ready::readable()
+    } else {
+        Ready::empty()
+    })
+}
+
+pub struct Events {
+    /// The Fuchsia selector only handles one event at a time, so there's no reason to
+    /// provide storage for multiple events.
+    event_opt: Option<Event>
+}
+
+impl Events {
+    pub fn with_capacity(_u: usize) -> Events { Events { event_opt: None } }
+    pub fn len(&self) -> usize {
+        if self.event_opt.is_some() { 1 } else { 0 }
+    }
+    pub fn capacity(&self) -> usize {
+        1
+    }
+    pub fn is_empty(&self) -> bool {
+        self.event_opt.is_none()
+    }
+    pub fn get(&self, idx: usize) -> Option<Event> {
+        if idx == 0 { self.event_opt } else { None }
+    }
+    pub fn push_event(&mut self, event: Event) {
+        assert!(::std::mem::replace(&mut self.event_opt, Some(event)).is_none(),
```

Comment:
Ah maybe not precisely like this? This'd also be a problem though if an I/O event comes in as well as a custom readiness event. The `push_event` [happens here](https://github.com/carllerche/mio/blob/a5415ba319e1bef09980f972caf4ffc9bf92d5b4/src/poll.rs#L2008) currently FWIW.

It should be ok to change this to a `Vec` right and just push at most one item during the call to `select`? It may be worth exploring writing a test for this as well, but that's ok to defer until later.